pkg/syncer/volume_health.go

```diff
@@ -77,58 +77,75 @@ func csiGetVolumeHealthStatus(ctx context.Context, k8sclient clientset.Interface
+				}
+			}
+		}
+		if !foundVol {
+			// Set volume health status as "Inaccessible" when PVC is not found in CNS.
+			// When a Datastore is removed from VC (like vSAN direct disk decommisson with noAction does), the CNS Volumes
+			// on that Datastore are eventually removed from CNS DB, but the PVCs still remain in the K8S cluster.
+			updateVolumeHealthStatus(ctx, k8sclient, pvc, common.VolHealthStatusInaccessible)
+		}
+	}
+	log.Infof("GetVolumeHealthStatus: end")
+}
 
-		if pvc, ok := volumeHandleToPvcMap[vol.VolumeId.Id]; ok {
-			log.Debugf("csiGetVolumeHealthStatus: Found pvc %q for volume %q", pvc, vol.VolumeId.Id)
+func updateVolumeHealthStatus(ctx context.Context, k8sclient clientset.Interface, pvc *v1.PersistentVolumeClaim, volHealthStatus string) {
+	log := logger.GetLogger(ctx)
+	log.Infof("updateVolumeHealthStatus: start")
 
-			// only update PVC health annotation if the HealthStatus of volume is not "unknown"
-			if vol.HealthStatus != string(pbmtypes.PbmHealthStatusForEntityUnknown) {
-				volHealthStatus, err := common.ConvertVolumeHealthStatus(vol.HealthStatus)
-				if err != nil {
-					log.Errorf("csiGetVolumeHealthStatus: invalid health status %q for volume %q", vol.HealthStatus, vol.VolumeId.Id)
-				}
-				val, found := pvc.Annotations[annVolumeHealth]
-				_, foundAnnHealthTS := pvc.Annotations[annVolumeHealthTS]
-				if !found || val != volHealthStatus || !foundAnnHealthTS {
-					// VolumeHealth annotation on pvc is changed, set it to new value
-					log.Debugf("csiGetVolumeHealthStatus: update volume health annotation for pvc %s/%s from old value %s to new value %s",
-						pvc.Namespace, pvc.Name, val, volHealthStatus)
-					metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annVolumeHealth, volHealthStatus)
-					metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annVolumeHealthTS, time.Now().Format(time.UnixDate))
-					log.Infof("csiGetVolumeHealthStatus: set annotation for health to %s at time %s for pvc %s/%s", volHealthStatus, time.Now().Format(time.UnixDate), pvc.Namespace, pvc.Name)
-					_, err := k8sclient.CoreV1().PersistentVolumeClaims(pvc.Namespace).Update(ctx, pvc, metav1.UpdateOptions{})
+	val, found := pvc.Annotations[annVolumeHealth]
+	_, foundAnnHealthTS := pvc.Annotations[annVolumeHealthTS]
+	if !found || val != volHealthStatus || !foundAnnHealthTS {
+		// VolumeHealth annotation on pvc is changed, set it to new value
+		log.Debugf("updateVolumeHealthStatus: update volume health annotation for pvc %s/%s from old value %s to new value %s",
+			pvc.Namespace, pvc.Name, val, volHealthStatus)
+		metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annVolumeHealth, volHealthStatus)
+		metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annVolumeHealthTS, time.Now().Format(time.UnixDate))
+		log.Infof("updateVolumeHealthStatus: set annotation for health to %s at time %s for pvc %s/%s", volHealthStatus, time.Now().Format(time.UnixDate), pvc.Namespace, pvc.Name)
```

Comment:
It is better to capture the timestamp in a variable before setting it in annotation and then logging it. This will give us the exact timestamp in the logs to help debugging.